public/app/features/dashboard/components/Inspector/InspectJSONTab.tsx

```diff
@@ -0,0 +1,207 @@
+      return this.props.panel.getSaveModel();
+    }
+
+    return { note: 'Unknown Object', show };
+  };
+
+  getClipboardText = () => {
+    const { show } = this.state;
+    const obj = this.getJSONObject(show);
+    return JSON.stringify(obj, null, 2);
+  };
+
+  onClipboardCopied = () => {
+    appEvents.emit(AppEvents.alertSuccess, ['Content copied to clipboard']);
+    alert('TODO... the notice is behind the inspector!');
+  };
+
+  onApplyPanelModel = () => {
+    const { panel, dashboard } = this.props;
+    try {
+      if (!dashboard.meta.canEdit) {
+        appEvents.emit(AppEvents.alertError, ['Unable to Apply']);
+      } else {
+        const newPanel = JSON.parse(this.state.text);
+        replacePanel(dashboard, newPanel, panel);
+        appEvents.emit(AppEvents.alertSuccess, ['Applied Changes']);
+      }
+    } catch (err) {
+      console.log('Error applyign updates', err);
+      appEvents.emit(AppEvents.alertError, ['Invalid JSON Text']);
+    }
+
+    // Close the inspector
+    getLocationSrv().update({
+      query: { inspect: null, tab: null },
+      partial: true,
+    });
+  };
+
+  render() {
+    const { dashboard } = this.props;
+    const { show } = this.state;
+    const selected = options.find(v => v.value === show);
+    const isPanelJSON = show === ShowContent.PanelJSON;
+    const canEdit = dashboard.meta.canEdit;
+
+    const styles = getStyles();
+
+    return (
+      <div className={styles.wrap}>
+        <Select options={options} value={selected} onChange={this.onSelectChanged} />
+        {selected.editText && canEdit ? (
+          <Forms.TextArea
+            spellCheck={false}
+            value={this.state.text}
+            onChange={this.onTextChanged}
+            className={styles.editor}
+            rows={20}
+          />
+        ) : (
+          <CustomScrollbar className={styles.content}>
+            <JSONFormatter json={this.getJSONObject(show)} open={4} />
+          </CustomScrollbar>
+        )}
+        <div>
+          {isPanelJSON && canEdit && (
+            <>
+              <Button onClick={this.onApplyPanelModel}>Apply</Button>&nbsp;
+            </>
+          )}
+          <ClipboardButton variant="secondary" getText={this.getClipboardText} onClipboardCopy={this.onClipboardCopied}>
+            Copy to Clipboard
+          </ClipboardButton>
+        </div>
+      </div>
+    );
+  }
+}
+
+function getSaveModelJSON(panel: PanelModel): string {
+  return JSON.stringify(panel.getSaveModel(), null, 2);
+}
+
+const getStyles = stylesFactory(() => {
+  return {
+    wrap: css`
+      display: flex;
+      flex-direction: column;
+      height: 100%;
+      width: 100%;
+    `,
+    content: css`
+      flex: 1;
+      border: 0px solid green;
+    `,
+    editor: css`
+      flex: 1; /* ???????? */
+      height: 100%;
+      border: 1px solid #222;
+      background: inherit;
+      color: inherit;
+      width: 100%;
+      margin-bottom: 5px;
+      font-family: monospace;
```

Comment:
TODO: this really should be a slate editor, not TextArea.

also... would be nice if adding `<Forms.TextArea>` to a component did not require adding special CSS (currently it shows up as ugly white box)

Reply:
> (currently it shows up as ugly white box)

it does? Seems to work fine in the DashboardImportPage

Reply:
It's due to a bug in TextArea component, if you pass it className it overrides all the internal styles, it should do  {...props} className={cx(styles.textarea, props.className)}  that way external passed className can be combined with internall styles

Reply:
Fixed in #23367